### ISSUE TITLE
Padding target strings with the SOS token

### ISSUE THREAD

**Issue by atliSig** [Alexander-H-Liu/End-to-end-ASR-Pytorch#22]
I noticed that you use the same key, 0, for both indicating start of sentence (SOS) 
`table = {'<sos>':0,'<eos>':1}`

as well as padding the targets
```
new_y = np.zeros((len(y),max_len),dtype=int)
for idx,label_seq in enumerate(y):
     new_y[idx,:len(label_seq)] = np.array(label_seq)
```

Meaning that a short sentence in a batch, e.g. "hello, world" could be encoded as
```
<sos>, 11, 8, 15, 15, 18, 48, 49, 24, 18, 20, 15,  7, <eos>, <sos>, <sos>, <sos>, <sos>, <sos>
```
depending on the mapping and the maximum length of the batch.

I haven't really thought about the consequences of this but wouldn't it be better to add some new pad token for targets?

**Comment by atliSig** [comment]
Another related thing:
The number of decoder steps passed on to the Speller is determined by this line `solver.py`:
```ans_len = int(torch.max(torch.sum(y!=0,dim=-1)))```
but to me it seems obvious that the number of decoder steps should just equal the second dimension of y (after squeezing) since the batch has already been padded with the aforementioned zeros to the maximum size of the batch.

**Comment by Alexander-H-Liu** [comment]
@atliSig 

- Regarding the padding token
Yes, I'm using 0 as both <sos> and padding token (in fact, also <blank> for CTC)
This is fine because I'm also setting 
``` torch.nn.CrossEntropyLoss(ignore_index=0, reduction='none')```
that is, 0 will not be considered as the target and the (attention based) encoder will never learn to output 0 no matter we treat 0 as <sos> or padding token 

- Regarding ```ans_len```
You're right, this is a redundant line, I'll remove it in the next version

Thank you!
